packages/hop-node/src/1inch/1inch.ts

```diff
@@ -0,0 +1,347 @@
+    })
+    const result = await this.getJson(url)
+
+    if (!result.data) {
+      logger.error(result)
+      throw new Error('expected tx data')
+    }
+
+    const { data, to, value } = result
+
+    return {
+      data,
+      to,
+      value
+    }
+  }
+
+  async getSwapTx (params: SwapParams) {
+    const { fromTokenAddress, toTokenAddress, fromAddress, amount, slippage, destReceiver } = params
+    if (!this.chainId) {
+      throw new Error('chainId is required')
+    }
+    if (!fromTokenAddress) {
+      throw new Error('fromTokenAddrss is required')
+    }
+    if (!toTokenAddress) {
+      throw new Error('toTokenAddress is required')
+    }
+    if (!fromAddress) {
+      throw new Error('fromAddress is required')
+    }
+    if (!amount) {
+      throw new Error('amount is required')
+    }
+    if (!slippage) {
+      throw new Error('slippage is required')
+    }
+
+    const url = this.constructUrl('/swap', {
+      fromTokenAddress,
+      toTokenAddress,
+      amount,
+      fromAddress,
+      slippage,
+      destReceiver
+    })
+    const result = await this.getJson(url)
+    if (!result.tx) {
+      logger.error(result)
+      throw new Error('expected tx data')
+    }
+
+    const { data, to, value } = result.tx
+
+    return {
+      data,
+      to,
+      value
+    }
+  }
+}
+
+type SwapInput = {
+  chain: string
+  fromToken: string
+  toToken: string
+  amount: number
+  max: boolean
+  slippage: number
+  recipient: string
+  dryMode: boolean
+}
+
+export async function swap (input: SwapInput) {
+  let {
+    chain,
+    fromToken,
+    toToken,
+    amount: formattedAmount,
+    max,
+    slippage = 0.1,
+    recipient,
+    dryMode
+  } = input
+  if (slippage > 1) {
+    throw new Error('slippage parameter is too high')
+  }
+
+  if (slippage <= 0) {
+    throw new Error('slippage must be higher than 0')
+  }
+
+  const wallet = wallets.get(chain)
+  const oneInch = new OneInch(chain)
+
+  const walletAddress = await wallet.getAddress()
+  let amount: string
+  const isFromNative = nativeChainTokens[chain] === fromToken
+  const isToNative = nativeChainTokens[chain] === toToken
+  if (max) {
+    if (isFromNative) {
+      amount = (await wallet.getBalance()).toString()
+      formattedAmount = Number(formatUnits(amount.toString(), getTokenDecimals(fromToken)))
+    } else {
+      const tokenContracts = contracts.get(fromToken, chain)
+      if (!tokenContracts) {
+        throw new Error('token contracts not found')
+      }
+      let token: Token
+      if (chain === Chain.Ethereum) {
+        token = new Token(tokenContracts.l1CanonicalToken)
+      } else {
+        token = new Token(tokenContracts.l2CanonicalToken)
+      }
+
+      amount = (await token.getBalance()).toString()
+      formattedAmount = Number(formatUnits(amount.toString(), getTokenDecimals(fromToken)))
+    }
+  } else {
+    amount = parseUnits(formattedAmount.toString(), getTokenDecimals(fromToken)).toString()
```

Comment:
We handle from native vs. token explicitly in the `(max)` case. Do we need to do the same here? In other words, does `getTokenDecimals()` work for ETH? I believe the answer is [yes](https://github.com/hop-protocol/hop/blob/749dcb013c22b666b3ef38bf7af525b1d1f4a922/packages/core/src/metadata/tokens.ts#L16-L20), but want to confirm.

Reply:
yes it works with eth symbol